### PR TITLE
Simplify linking, update dependencies

### PR DESCRIPTION
**Pull request by julianoes**
On Windows, the mavsdk_server unit tests don't link properly when building with shared libs. For whatever reason the transitive dependencies didn't get tracked from the protobu/grpc components all the way into the mavsdk_server unit tests.

I suspect that either some symbols get stripped too early, or it has to do with the export visibility in MSVC.

I couldn't quite get to the bottom of it but I realized that we could just simplify it by avoiding the little libraries and just adding all sources to the mavsdk_server library in the first place.

Let's see if it works...

Edit: same problem...

### CODE REVIEW

**Comment by julianoes** [comment]
The problem persists, this is what the linking error looks like:

```
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::VelocityBody::`vftable'" (??_7VelocityBody@telemetry@rpc@mavsdk@@6B@) ref
erenced in function "public: __cdecl mavsdk::rpc::telemetry::VelocityBody::VelocityBody(void)" (??0VelocityBody@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Users\Julian Oes\src\MAVSDK\build
\default\src\mavsdk_server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::PositionBody::`vftable'" (??_7PositionBody@telemetry@rpc@mavsdk@@6B@) ref
erenced in function "public: __cdecl mavsdk::rpc::telemetry::PositionBody::PositionBody(void)" (??0PositionBody@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Users\Julian Oes\src\MAVSDK\build
\default\src\mavsdk_server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::Odometry::`vftable'" (??_7Odometry@telemetry@rpc@mavsdk@@6B@) referenced
in function "public: __cdecl mavsdk::rpc::telemetry::Odometry::Odometry(void)" (??0Odometry@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Users\Julian Oes\src\MAVSDK\build\default\src\mavsdk_
server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::DistanceSensor::`vftable'" (??_7DistanceSensor@telemetry@rpc@mavsdk@@6B@)
 referenced in function "public: __cdecl mavsdk::rpc::telemetry::DistanceSensor::DistanceSensor(void)" (??0DistanceSensor@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Users\Julian Oes\src\MA
VSDK\build\default\src\mavsdk_server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::ScaledPressure::`vftable'" (??_7ScaledPressure@telemetry@rpc@mavsdk@@6B@)
 referenced in function "public: __cdecl mavsdk::rpc::telemetry::ScaledPressure::ScaledPressure(void)" (??0ScaledPressure@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Users\Julian Oes\src\MA
VSDK\build\default\src\mavsdk_server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::PositionNed::`vftable'" (??_7PositionNed@telemetry@rpc@mavsdk@@6B@) refer
enced in function "public: __cdecl mavsdk::rpc::telemetry::PositionNed::PositionNed(void)" (??0PositionNed@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Users\Julian Oes\src\MAVSDK\build\defa
ult\src\mavsdk_server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::VelocityNed::`vftable'" (??_7VelocityNed@telemetry@rpc@mavsdk@@6B@) refer
enced in function "public: __cdecl mavsdk::rpc::telemetry::VelocityNed::VelocityNed(void)" (??0VelocityNed@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Users\Julian Oes\src\MAVSDK\build\defa
ult\src\mavsdk_server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::PositionVelocityNed::`vftable'" (??_7PositionVelocityNed@telemetry@rpc@ma
vsdk@@6B@) referenced in function "public: __cdecl mavsdk::rpc::telemetry::PositionVelocityNed::PositionVelocityNed(void)" (??0PositionVelocityNed@telemetry@rpc@mavsdk@@QEAA@XZ) [C
:\Users\Julian Oes\src\MAVSDK\build\default\src\mavsdk_server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::GroundTruth::`vftable'" (??_7GroundTruth@telemetry@rpc@mavsdk@@6B@) refer
enced in function "public: __cdecl mavsdk::rpc::telemetry::GroundTruth::GroundTruth(void)" (??0GroundTruth@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Users\Julian Oes\src\MAVSDK\build\defa
ult\src\mavsdk_server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::FixedwingMetrics::`vftable'" (??_7FixedwingMetrics@telemetry@rpc@mavsdk@@
6B@) referenced in function "public: __cdecl mavsdk::rpc::telemetry::FixedwingMetrics::FixedwingMetrics(void)" (??0FixedwingMetrics@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Users\Julian
Oes\src\MAVSDK\build\default\src\mavsdk_server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::AccelerationFrd::`vftable'" (??_7AccelerationFrd@telemetry@rpc@mavsdk@@6B
@) referenced in function "public: __cdecl mavsdk::rpc::telemetry::AccelerationFrd::AccelerationFrd(void)" (??0AccelerationFrd@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Users\Julian Oes\s
rc\MAVSDK\build\default\src\mavsdk_server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::AngularVelocityFrd::`vftable'" (??_7AngularVelocityFrd@telemetry@rpc@mavs
dk@@6B@) referenced in function "public: __cdecl mavsdk::rpc::telemetry::AngularVelocityFrd::AngularVelocityFrd(void)" (??0AngularVelocityFrd@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Use
rs\Julian Oes\src\MAVSDK\build\default\src\mavsdk_server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::MagneticFieldFrd::`vftable'" (??_7MagneticFieldFrd@telemetry@rpc@mavsdk@@
6B@) referenced in function "public: __cdecl mavsdk::rpc::telemetry::MagneticFieldFrd::MagneticFieldFrd(void)" (??0MagneticFieldFrd@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Users\Julian
Oes\src\MAVSDK\build\default\src\mavsdk_server\test\unit_tests_mavsdk_server.vcxproj]
telemetry_service_impl_test.obj : error LNK2019: unresolved external symbol "const mavsdk::rpc::telemetry::Imu::`vftable'" (??_7Imu@telemetry@rpc@mavsdk@@6B@) referenced in functio
n "public: __cdecl mavsdk::rpc::telemetry::Imu::Imu(void)" (??0Imu@telemetry@rpc@mavsdk@@QEAA@XZ) [C:\Users\Julian Oes\src\MAVSDK\build\default\src\mavsdk_server\test\unit_tests_ma
```

**Review by JonasVautherin — APPROVED** [review]
Makes sense. I think it was made like this such that it would trigger the build when the proto files changed or something like this. But then we moved the generation out of CMake and versioned the generated files.

So this is simpler indeed!

**Comment by JonasVautherin** [comment]
Not completely sure what the latest changes imply. Do they solve the Windows problem?

**Comment by julianoes** [comment]
Nope, nothing solves the Windows problem :disappointed:

File: third_party/openssl/CMakeLists.txt [hunk 1]
```diff
@@ -36,7 +36,7 @@ if(ANDROID)
         PREFIX openssl
         PATCH_COMMAND patch -p 0 < ${PROJECT_SOURCE_DIR}/dockcross-android.patch
         CONFIGURE_COMMAND ${CMAKE_COMMAND} -E env ANDROID_NDK=$ENV{CROSS_ROOT} <SOURCE_DIR>/Configure --sysroot=$ENV{CROSS_ROOT}/sysroot ${OPENSSL_PLATFORM} --prefix=${CMAKE_INSTALL_PREFIX} ${OPENSSL_BUILD_TYPE} no-shared
-        BUILD_COMMAND make
+        BUILD_COMMAND make install_sw
```
Comment (JonasVautherin) [inline third_party/openssl/CMakeLists.txt:39]: Sounds like an `INSTALL_COMMAND`?

File: third_party/zlib/CMakeLists.txt [hunk 2]
```diff
@@ -27,8 +27,8 @@ endforeach()
 ExternalProject_add(
     zlib
     GIT_REPOSITORY https://github.com/madler/zlib
-    GIT_TAG v1.2.12
+    GIT_TAG v1.2.13
     PREFIX zlib
-    PATCH_COMMAND git checkout . && git apply ${PROJECT_SOURCE_DIR}/build_shared_libs.patch
+    PATCH_COMMAND git checkout . && git apply ${PROJECT_SOURCE_DIR}/build_shared_libs_and_cmake_version.patch
```
Comment (JonasVautherin) [inline third_party/zlib/CMakeLists.txt:32]: Can we split that in two PATCH_COMMAND lines? They seem unrelated :thinking:

File: third_party/grpc/cmake_install_fix.patch [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
++++ b/CMakeLists.txt
+@@ -26091,10 +26091,12 @@ if(gRPC_INSTALL)
+     DESTINATION ${gRPC_INSTALL_CMAKEDIR}
+     NAMESPACE gRPC::
+   )
+-  install(EXPORT gRPCPluginTargets
+-    DESTINATION ${gRPC_INSTALL_CMAKEDIR}
+-    NAMESPACE gRPC::
+-  )
++  if (gRPCPluginTargets AND gRPC_BUILD_CODEGEN)
```
Comment (JonasVautherin) [inline third_party/grpc/cmake_install_fix.patch:23]: Is that a bug in gRPC upstream?

**Review by JonasVautherin — CHANGES_REQUESTED** [review]

File: third_party/openssl/CMakeLists.txt (hunk 1 above)
Comment (julianoes) [reply third_party/openssl/CMakeLists.txt:39]: Seems to work though.

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.13)
+cmake_minimum_required(VERSION 3.13...3.22)
```
[the remark this answers is not in the document]
Comment (julianoes) [reply CMakeLists.txt:1]: It's just a max version that it has been tested with and policies set. I added it because there was a warning about it but it actually came from the zlib CMakeLists.txt where the minimum version was 2.4.

File: third_party/grpc/cmake_install_fix.patch (hunk 3 above)
Comment (julianoes) [reply third_party/grpc/cmake_install_fix.patch:23]: I think so and it has supposedly been fixed but either the fix didn't go into v1.55.1 or it doesn't fully fix it for us for some reason:
https://github.com/grpc/grpc/issues/32885

File: third_party/zlib/CMakeLists.txt (hunk 2 above)
Comment (julianoes) [reply third_party/zlib/CMakeLists.txt:32]: Sure. I just thought putting it into one patch is easier.

File: third_party/openssl/CMakeLists.txt (hunk 1 above)
Comment (JonasVautherin) [reply third_party/openssl/CMakeLists.txt:39]: Sure, we could write `./configure && make && make install` in the CONFIGURE_COMMAND, and CMake wouldn't know about it. But we would :grin:.

It's nice if the log output says "starting install step" before running the installation step, I think :innocent:

File: third_party/zlib/CMakeLists.txt (hunk 2 above)
Comment (JonasVautherin) [reply third_party/zlib/CMakeLists.txt:32]: Ok, as you want. I see patches like commits. It's easier for me to read a commit that does one thing than to try to understand which line corresponds to what when reading the commit.

But if you like one patch better, technically it works :+1:.

File: third_party/openssl/CMakeLists.txt (hunk 1 above)
Comment (julianoes) [reply third_party/openssl/CMakeLists.txt:39]: Let me try.

File: third_party/zlib/CMakeLists.txt [hunk 5]
```diff
@@ -27,8 +27,8 @@ endforeach()
 ExternalProject_add(
     zlib
     GIT_REPOSITORY https://github.com/madler/zlib
-    GIT_TAG v1.2.12
+    GIT_TAG v1.2.13
     PREFIX zlib
-    PATCH_COMMAND git checkout . && git apply ${PROJECT_SOURCE_DIR}/build_shared_libs.patch
+    PATCH_COMMAND git checkout . && git apply ${PROJECT_SOURCE_DIR}/build_shared_libs.patch && git apply ${PROJECT_SOURCE_DIR}/bump_cmake_min_version.patch
```
Comment (JonasVautherin) [inline third_party/zlib/CMakeLists.txt:32]: Note (but it does not matter much). You can add as many `COMMAND` as you want between the "key" commands. So this should work:

```cmake
PATCH_COMMAND git checkout . && git apply ${PROJECT_SOURCE_DIR}/build_shared_libs.patch
COMMAND git apply ${PROJECT_SOURCE_DIR}/bump_cmake_min_version.patch
```

See the [Miscellaneous Options](https://cmake.org/cmake/help/latest/module/ExternalProject.html) in the docs:

> Any of the other ..._COMMAND options can have additional commands appended to them by following them with as many COMMAND ... options as needed

**Review by JonasVautherin — APPROVED** [review]

File: third_party/zlib/CMakeLists.txt (hunk 5 above)
Comment (julianoes) [reply third_party/zlib/CMakeLists.txt:32]: Yer, on the other hand I like having all of the 3 commands together wrapped as "the" patch command.

**Comment by julianoes** [comment]
I think I want to clean up the commits a bit before merging.
